server/audit_math/bravo.py

```diff
@@ -34,45 +35,59 @@ def get_expected_sample_sizes(
                         }
 
     Output:
-        expected sample size - the expected sample size for the contest
+        the expected sample size or ASN for the contest
     """
+    if is_tie(contest):
+        raise ValueError("Cannot compute ASN for a tied contest")
 
-    margin = contest.margins
-    p_w = Decimal("inf")
-    s_w = Decimal(0.0)
-    p_l = Decimal(0.0)
-    # Get smallest p_w - p_l
-    for winner in margin["winners"]:
-        if margin["winners"][winner]["p_w"] < p_w:
-            p_w = Decimal(margin["winners"][winner]["p_w"])
-
-    # If there aren't any losers,
-    if not margin["losers"]:
-        return -1
-
-    for loser in margin["losers"]:
-        if margin["losers"][loser]["p_l"] > p_l:
-            p_l = Decimal(margin["losers"][loser]["p_l"])
-
-    s_w = p_w / (p_w + p_l)
-
-    if p_w == 1.0 or contest.num_winners >= len(contest.candidates):
-        # Handle single-candidate or crazy landslides
-        return -1
-    elif p_w == p_l:
-        return contest.ballots
-    else:
-        z_w = (2 * s_w).ln()
-        z_l = (2 - 2 * s_w).ln()
-
-        T = Decimal(min(get_test_statistics(contest.margins, sample_results).values()))
+    winners = contest.margins["winners"]
+    losers = contest.margins["losers"]
 
-        weighted_alpha = (Decimal(1.0) / alpha) / T
-        return int(
-            (
+    if not losers:
+        raise ValueError("Cannot compute ASN for a contest with no losers")
+
+    T = get_test_statistics(contest.margins, cumulative_sample_results)
+
+    sample_size = 0
+    probability_of_completion = None
+    for winner_name, winner_stats in winners.items():
+        for loser_name, loser_stats in losers.items():
+            weighted_alpha = (Decimal(1) / alpha) / T[(winner_name, loser_name)]
+            p_w = Decimal(winner_stats["p_w"])
+            p_l = Decimal(loser_stats["p_l"])
+            if p_l == 0:
+                # p_l is 0 --> s_w is 1 --> z_l is -Infinity
+                # Skip this pair to avoid invalid operation errors
+                continue
+            s_w = p_w / (p_w + p_l)
+            z_w = (2 * s_w).ln()
+            z_l = (2 - 2 * s_w).ln()
+
+            possible_sample_size = math.ceil(
                 (weighted_alpha.ln() + (z_w / Decimal(2))) / (p_w * z_w + p_l * z_l)
-            ).quantize(Decimal(1), rounding=ROUND_CEILING)
-        )
+            )
```

Comment:
Subtle tweak here. We were previously using `.quantize(Decimal(1), rounding=ROUND_CEILING)` to round up to the nearest integer. I now just use `math.ceil` because 1) it's simpler to understand and 2) `.quantize` sometimes throws invalid operation errors due to precision issues ([reference](https://stackoverflow.com/questions/31795109/decimal-quantize-raises-invalidoperation))

Reply:
I think this may have to change back. We were using the Decimal class specifically because of it's improved precision. This discussion is **old** so it might be worth discussing again, but here's the PR that introduced the Decimal module: https://github.com/votingworks/arlo/pull/829

Reply:
Hmm, thinking about this some more, maybe this is okay. Since Decimal was being used to handle an integer value, I doubt we'll get into a situation where that's an issue...all the floats are still being treated as Decimals, so I think this is fine.

Reply:
Yup my thinking exactly. I see the importance of using `Decimal` for the intermediate calculations, but for this last step, where we're rounding to an integer, it's not as relevant